### PR TITLE
Crypto: define external crypto backend API and implement client

### PR DESCRIPTION
**Pull request by reinkrul**
Closes https://github.com/nuts-foundation/nuts-node/issues/1837

TODO:
- [x] Documentation
- [ ] End-to-end test (see https://github.com/nuts-foundation/nuts-go-e2e-test/pull/57)

### CODE REVIEW

**Comment by stevenvegt** [comment]
Tests are broken on master, fixed in #1856

File: docs/pages/development/6-external-secret-store.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+Implementing a proxy or building a custom secret store
+******************************************************
+
+A proxy or secret store should implement the Nuts Secret store API specification. This OpenAPI specification is available on `GitHub <https://raw.githubusercontent.com/Nuts-foundation/Nuts-node/master/docs/_static/crypto/Nuts-storage-api-v1.yaml>`__.
+
+Consider developing your implementation under an open source license and publish it on `GitHub <https://github.com>`__ so that other parties can use it.
```
Comment (woutslakhorst) [inline docs/pages/development/6-external-secret-store.rst:20]: Gitlab is not allowed?

File: docs/pages/development/6-external-secret-store.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      storage: external
+      external:
+        address: https://localhost:8210
+
+Limitations
+***********
+
+The API has a few limitations:
+
+- It does not support authentication. This means that the proxy should be secured in such a way that only the Nuts node can access the API.
```
Comment (woutslakhorst) [inline docs/pages/development/6-external-secret-store.rst:39]: (yet)

File: docs/pages/deployment/monitoring.rst [hunk 3]
```diff
@@ -7,7 +7,7 @@ Health checks
 *************
 
 An endpoint is provided to perform health checks on the Nuts node.
-It reports in a format compatible with :ref:`Spring Boot's Health Actuator <https://docs.spring.io/spring-boot/docs/2.0.x/actuator-api/html/#health>`.
+It reports in a format compatible with `Spring Boot's Health Actuator <https://docs.spring.io/spring-boot/docs/2.0.x/actuator-api/html/#health>`__.
```
Comment (woutslakhorst) [inline docs/pages/deployment/monitoring.rst:10]: Single underscore at end

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
Generate README

File: core/status/engine.go [hunk 4]
```diff
@@ -117,6 +119,13 @@ func (s *status) checkHealth(ctx echo.Context) error {
 	result := s.doCheckHealth()
 	responseCode := 200
 	if result.Status != core.HealthStatusUp {
+		var failures []string
+		for component, componentHealth := range result.Details.(map[string]core.Health) {
+			if componentHealth.Status != core.HealthStatusUp {
+				failures = append(failures, fmt.Sprintf(" - %s: %v", component, componentHealth.Details))
+			}
+		}
+		logrus.Warnf("Health check status is not UP, failing components:\n%s", strings.Join(failures, "\n"))
```
[the remark this answers is not in the document]
Comment (stevenvegt) [reply core/status/engine.go:128]: If your node is down for a few hours, you do now want to scroll through multiple hours of logs to find this single message. This is a severe problem and should not occur often. I even think we should change it from Warn to Error.

File: crypto/storage/external/client.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	response, err := c.httpClient.HealthCheckWithResponse(context.Background())
+	if err != nil {
+		results[StorageType] = core.Health{Status: core.HealthStatusDown, Details: fmt.Errorf("unable to connect to storage server: %w", err).Error()}
+		return results
+	}
+
+	switch response.StatusCode() {
+	case http.StatusOK:
+		results[StorageType] = core.Health{Status: core.HealthStatusUp}
+	case http.StatusServiceUnavailable:
```
[the remark this answers is not in the document]
Comment (stevenvegt) [reply crypto/storage/external/client.go:59]: Ok, changed it to: if statusCode != HttpOK then DOWN

File: docs/pages/deployment/monitoring.rst (hunk 3 above)
Comment (stevenvegt) [reply docs/pages/deployment/monitoring.rst:10]: Only if you want to refer to them see: https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#anonymous-hyperlinks and https://stackoverflow.com/questions/27420317/restructured-text-rst-http-links-underscore-vs-use

File: core/status/engine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply core/status/engine.go:128]: It'll be in your healthcheck output. This should be picked up by other services which will spam you by sms/app. There's no need to fill the logs with the same message over and over again.

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
Still missing updated README

**Review by woutslakhorst — APPROVED** [review]

File: core/status/engine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (stevenvegt) [reply core/status/engine.go:128]: As decided by committee, lets keep this in here and let the log service deal with it.

**Pull request merged**